baselines/niid_bench/README.md

```diff
@@ -0,0 +1,131 @@
+
+****Contributors:**** Aashish Kolluri, PhD Candidate, National University of Singapore
+
+
+## Experimental Setup
+
+****Task:**** Image classification
+
+****Model:**** This directory implements CNNs as mentioned in the paper (Section V, paragraph 1). Specifically, the CNNs have two 2D convolutional layers with 6 and 16 output channels, kernel size 5, and stride 1.
+
+****Dataset:**** This directory has three image classification datasets that are used in the baseline, MNIST, CIFAR10, and Fashion-mnist. Further, five different data-splitting strategies are used including iid and four non-iid strategies based on label skewness. In the first non-iid strategy, for each label the data is split based on proportions sampled from a dirichlet distribution (with parameter 0.5). In the three remaining strategies, each client gets data from randomly chosen #C labels where #C is 1, 2, or 3. For the clients that are supposed to receive data from the same label the data is equally split between them. The baseline considers 10 clients. The following table shows all dataset and data splitting configurations.
+
+| Dataset | #classes | #partitions | partitioning method | partition settings |
+| :------ | :---: | :---: | :---: | :---: |
+| CIFAR10 | 10 | 10 | IID | NA |
+| CIFAR10 | 10 | 10 | dirichlet | distribution parameter 0.5 |
+| CIFAR10 | 10 | 10 | sort and partition | 1 label per client |
+| CIFAR10 | 10 | 10 | sort and partition | 2 labels per client |
+| CIFAR10 | 10 | 10 | sort and partition | 3 labels per client |
+| MNIST | 10 | 10 | IID | NA |
+| MNIST | 10 | 10 | dirichlet | distribution parameter 0.5 |
+| MNIST | 10 | 10 | sort and partition | 1 label per client |
+| MNIST | 10 | 10 | sort and partition | 2 labels per client |
+| MNIST | 10 | 10 | sort and partition | 3 labels per client |
+| FMNIST | 10 | 10 | IID | NA |
+| FMNIST | 10 | 10 | dirichlet | distribution parameter 0.5 |
+| FMNIST | 10 | 10 | sort and partition | 1 label per client |
+| FMNIST | 10 | 10 | sort and partition | 2 labels per client |
+| FMNIST | 10 | 10 | sort and partition | 3 labels per client |
+
+
+****Training Hyperparameters:**** There are four FL algorithms and they have many common hyperparameters and few different ones. The following table shows the common hyperparameters and their default values.
+
+| Description | Default Value |
+| ----------- | ----- |
+| total clients | 10 |
+| clients per round | 10 |
+| number of rounds | 50 |
+| number of local epochs | 10 |
+| client resources | {'num_cpus': 4.0, 'num_gpus': 0.0 }|
+| dataset name | cifar10 
+| data partition | Dirichlet (0.5) |
+| batch size | 64 |
+| momentum for SGD | 0.9 |
+
+For FedProx algorithm the proximal parameter is tuned from values {0.001, 0.01, 0.1, 1.0} in the experiments. The default value is 0.01. 
+
+
+## Environment Setup
+
+```bash
+# Setup the base poetry enviroment from the niid_bench directory
+poetry install
+
+# Start the shell
+poetry shell
```

Comment:
```suggestion
# Set python version
pyenv local 3.10.6
# Tell poetry to use python 3.10
poetry env use 3.10.6
# Now install the environment
poetry install
# Start the shell
poetry shell
```

Reply:
We require now all baselines to use python3.10+. A change related to this is also shown in the `pyproject.toml`.